### PR TITLE
Optimize hash strategy

### PR DESCRIPTION
**Pull request by ilackarms**
reduces CPU overhead in gloo containers
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/706

### CODE REVIEW

File: test/benchmark/snapshot_benchmark_test.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		b.RecordValue("Runtime per generated call in µ seconds", float64(int64(runtime2)/times)/1e3)
+
+	}, 10)
+})
+
+type pair struct {
+	k, v string
+}
+
+func Hashi(us *v1.Upstream) uint64 {
```
Comment (ilackarms) [inline test/benchmark/snapshot_benchmark_test.go]: testing a replacement function for upstream hashing
not currently used

File: projects/gloo/pkg/upstreams/hybrid_client.go [hunk 2]
```diff
@@ -140,17 +141,23 @@ func (c *hybridUpstreamClient) Watch(namespace string, opts clients.WatchOpts) (
...
 		syncFunc := func() {
-			if current.hash() == previous.hash() {
+			currentHash := current.hash()
+			if currentHash == previousHash {
 				return
 			}
+			previousHash = currentHash
 			previous = current.clone()
 			toSend := current.clone()
 			upstreamsOut <- toSend.toList()
```
Comment (yuval-k) [inline projects/gloo/pkg/upstreams/hybrid_client.go]: do we want to select here, similar to how we do it in solo-kit?
and then only update the hash when successfully sent?

**Comment by ilackarms** [comment]
related to #1032

**Pull request merged**